### PR TITLE
Adziu/lesson 01 basics

### PR DESCRIPTION
**Pull request by RCanDo**
@adziu 
I hope everything ok.

### CODE REVIEW

**Comment by adziu** [comment]
We will be lax about the deadlines for a while, so no worries. We are also only tracking the progress of the tasks, that means linting and giving the right result. As for the suggestions, we expect that you use them in the next lessons.

So first off, you have a strange commit history here. You probably didn't start with the master and then merged it. You can get the history right with `git rebase origin/master`, and then `push -f fork adziu/lesson_01_basics` assuming that the `origin` and `fork` are aliases to original repo and your fork (which they probably are not, because if you cloned from your repo, then `origin` is your repo, unless you reconfigure it).

Look at the instructions again, some steps were added, to avoid that.

File: students/arkadiusz_kasprzyk/lesson_01_basics/school_desks.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    The second group has 21 students, so they can get by with no fewer than 11 desks.
+    11 desks is also enough for the third group of 22 students.
+    So we need 32 desks in total.
+'''
+
+a = int(input("The number of students in class A is: "))
+b = int(input("The number of students in class B is: "))
+c = int(input("The number of students in class C is: "))
+
+x = a // 2 + a%2
```
Comment (adziu) [inline students/arkadiusz_kasprzyk/lesson_01_basics/school_desks.py]: Linter should complain about missing spaces, it's a bad idea to hint at precedence (you might be wrong and a review won't catch it, if you give suggestions like that).

File: students/arkadiusz_kasprzyk/lesson_01_basics/apple_sharing.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
+'''
+title: apple_sharing
+author: arkadiusz.kasprzyk@tieto.com
+date: 2018-03-05
```
[the remark this answers is not in the document]
Comment (RCanDo) [reply students/arkadiusz_kasprzyk/lesson_01_basics/apple_sharing.py]: If the files were elements of one project then OK. But they are not - each is separate being.
I'd say that DRY principle is about code not about metainformations in comments or headings.
Of curse, it's rarely good to repeat yourself but sometimes it's vary useful; in case of headings - for me, it's harmless if only not too much of it. E.g. repeating title of a file allows to change the name of the file without worry that info about its content is lost forever. And so on.
The term "redundancy" is not always pejorative: information replicated is protected against vanishing. 
Obviously, it doesn't work in case of code, where RY makes it prone to bugs and less readable. No doubt about it.

File: students/arkadiusz_kasprzyk/lesson_01_basics/car_route.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    A car can cover distance of N kilometers per day.
+    How many days will it take to cover a route of length M kilometers?
+    '''
+)
+
+N = float(input("Number of km per day: "))
+M = float(input("Number of km to cover: "))
+
+f = m.floor(M/N)
+r = f + int(abs(M-f) > tol)
```
[the remark this answers is not in the document]
Comment (RCanDo) [reply students/arkadiusz_kasprzyk/lesson_01_basics/car_route.py]: It doesn't work because `-` sign disappeared: there should be `tol = 1e-9` not `1e9`. Just typo. 
I agree that `ceil` is better. Just fixation on using `floor`.
By the way, what's the general tactic to avoid usig tolerance?

File: students/arkadiusz_kasprzyk/lesson_01_basics/clock_face_1.py [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+
+mins = mins + secs // 60
+secs = secs % 60
+
+degs = degs + mins // 60
+mins = mins % 60
+
+
+print("Angle (in deg.min.sec) of the hour hand is {}° {}' {}''".format(degs, mins, secs))
```
[the remark this answers is not in the document]
Comment (RCanDo) [reply students/arkadiusz_kasprzyk/lesson_01_basics/clock_face_1.py]: It's really not clear what they mean. I've chosen more traditional way of noting angles (and little more chalenging) and got rapped... After all it works!

File: students/arkadiusz_kasprzyk/lesson_01_basics/total_cost.py [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+date: 2018-03-05
+description:
+    A cupcake costs A dollars and B cents.
+    Determine, how many dollars and cents should one pay for N cupcakes.
+    A program gets three numbers: A, B, N.
+    It should print two numbers: total cost in dollars and cents.
+'''
+
+import math as m
+tol = 1e9  ## tolerance
```
[the remark this answers is not in the document]
Comment (RCanDo) [reply students/arkadiusz_kasprzyk/lesson_01_basics/total_cost.py]: Ooops, copy-paste mistake. Didn't catch it...

File: students/arkadiusz_kasprzyk/lesson_01_basics/apple_sharing.py [hunk 6]
```diff
@@ -0,0 +1,32 @@
+'''
```
[the remark this answers is not in the document]
Comment (RCanDo) [reply students/arkadiusz_kasprzyk/lesson_01_basics/apple_sharing.py]: Pity, it's a bit easier to write `'` - no need to press Shift...

File: students/arkadiusz_kasprzyk/lesson_01_basics/clock_face_1.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (RCanDo) [reply students/arkadiusz_kasprzyk/lesson_01_basics/clock_face_1.py]: I haven't looked at test cases - it could have spoiled the fun of solving problem.

File: students/arkadiusz_kasprzyk/lesson_01_basics/car_route.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (adziu) [reply students/arkadiusz_kasprzyk/lesson_01_basics/car_route.py]: I'm no expert in numerical methods, but the general rule is that whenever floating points approach zero or infinity it gets unsafe: they may change sign, are unreliable when compared or cause big errors when multiplied. The idea is to keep the numbers close to the middle of the range. I just try to use libraries and do unit tests with some edge cases, because I rarely deal with that and only dig down deeper, if something goes wrong.

File: students/arkadiusz_kasprzyk/lesson_01_basics/apple_sharing.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (adziu) [reply students/arkadiusz_kasprzyk/lesson_01_basics/apple_sharing.py]: The redundancy is not pejorative, when computer does it ;)

File: students/arkadiusz_kasprzyk/lesson_01_basics/clock_face_1.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (adziu) [reply students/arkadiusz_kasprzyk/lesson_01_basics/clock_face_1.py]: Like I wrote in the introduction it's not personal. It's mostly about Python and program engineering. So from my experience in real world:

You get the requirements in written form or as unit tests (which is really great, but rarely happens) and your task is to pass, because that's the way to cooperate with other people on the project. Most of the time the requirements are unclear and you have to ask, you should rather be happy, if you get unit tests beforehand.

Also, most of the time the part you write is not a standalone solution, but rather a part of a solution and any superfluous input handling, prints or format mismatch will not be appreciated. To me it's "egal", because it's not real world. I'm just trying to pass that knowledge on.

**Comment by RCanDo** [comment]
@adziu 
Only most important corrections introduced.
I've left `input("Press enter to quit.")` because it allows to see results when running scripts directly in a console - without this window closes immediately. 
By the way, I really don't know what's going on while making pull request. For me it's all mess (I mean GitHub - some day I'll probably get used to it but my intuition doesn't really work here...). 
It's also not clear if master on my fork repo should be updated with master on the original repo. If so then every single lesson will contain all the previous. Is that OK? It seems more resonable to preserve master in its initial state, but what you really need?  Many questions...

**Comment by adziu** [comment]
Good enough, I think it'll be more productive if we continue with the next lessons. Replying to your concerns:
- the closing of the console window should be configurable,
- A git repository on the server is a graph of snapshots of code tree. Git repo on your local machine is that graph plus your working tree and a staging area for preparing new snapshots. Pull request here is a request to pull code snapshots from your fork into the graph of the original repository. If  you want, we can talk a bit what particular commands in git are doing.
- You should update the master and start the new branch of it. `git checkout master` `git fetch origin` `git pull --rebase origin master` `git checkout -b adziu/lesson_02_flow_control` and then you work on it,
- Your fork's master is not a concern, unless you want to make pull requests there and review the changes.

**Comment by adziu** [comment]
And as you can see the stickler is not working, because you are not rebased on upstream/master. I will ignore that. If you want to try the rebase I'll not merge that for a day yet.
